src/test/java/com/fasterxml/jackson/databind/cfg/CacheProviderTest.java

```diff
@@ -0,0 +1,179 @@
+            invoked = true;
+            return _cachedDeserializers.put(key, value);
+        }
+
+        @Override
+        public JsonDeserializer<Object> putIfAbsent(JavaType key, JsonDeserializer<Object> value) {
+            invoked = true;
+            return _cachedDeserializers.putIfAbsent(key, value);
+        }
+
+        @Override
+        public void clear() {
+            _cachedDeserializers.clear();
+        }
+    }
+    
+    static class CustomCacheProvider implements CacheProvider {
+        @Override
+        public LookupCache<JavaType, JsonDeserializer<Object>> forDeserializerCache(DeserializationConfig config) {
+            invoked = true;
+            return new SimpleTestCache(123);
+        }
+    }
+    
+    /*
+    /**********************************************************************
+    /* Unit tests
+    /**********************************************************************
+     */
+
+    // For checking CustomCacheProvider.forDeserializerCache() was actually used
+    private static boolean invoked = false;
+
+    // For checking the cache was actually created
+    private static final List<Map<JavaType, JsonDeserializer<Object>>> createdCaches = new ArrayList<>();
+    
+    @Before
+    public void setUp() {
+        createdCaches.clear();
+        invoked = false;
+    }
+
+    @Test
+    public void testDefaultCacheProviderConfigDeserializerCache() throws Exception
+    {
+        CacheProvider cacheProvider = DefaultCacheProvider.builder()
+                .maxDeserializerCacheSize(1234)
+                .build();
+        ObjectMapper mapper = JsonMapper.builder()
+                .cacheProvider(cacheProvider).build();
+
+        assertNotNull(mapper.readValue("{\"point\":24}", RandomBean.class));
+    }
+
+    @Test
+    public void testDefaultCacheProviderConfigDeserializerCacheSizeZero() throws Exception
+    {
+        CacheProvider cacheProvider = DefaultCacheProvider.builder()
+                .maxDeserializerCacheSize(0)
+                .build();
+        ObjectMapper mapper = JsonMapper.builder()
+                .cacheProvider(cacheProvider)
+                .build();
+
+        assertNotNull(mapper.readValue("{\"point\":24}", RandomBean.class));
+    }
+
+    @Test
+    public void testBuilderNullCheckingForDeserializerCacheConfig() throws Exception
+    {
+        try {
+            DefaultCacheProvider.builder()
+                    .maxDeserializerCacheSize(-1);
+            fail("Should not reach here");
+        } catch (IllegalArgumentException e) {
+            assertTrue(e.getMessage().contains("Cannot set maxDeserializerCacheSize to a negative value"));
+        }
+    }
+
+    @Test
+    public void testCustomCacheProviderConfig() throws Exception
+    {
+        ObjectMapper mapper = JsonMapper.builder()
+                .cacheProvider(new CustomCacheProvider())
+                .build();
+
+        assertTrue(invoked);
```

Comment:
I think it'd be better to make `CustomCacheProvider` itself stateful, possibly returning a shared instance of test `SimpleTestCache` and storing state in there -- relying on `static` in test class is bit tightly coupled.